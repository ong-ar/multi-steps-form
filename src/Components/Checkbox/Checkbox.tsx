import * as React from "react";

interface IProps {
  name?: string;
  value?: string;
  disabled?: boolean;
  checked?: boolean;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const Checkbox: React.SFC<IProps> = ({
  name,
  value,
  disabled,
  checked,
  onChange,
  children
}) => {
  const inputStyle: React.CSSProperties = {
    marginLeft: -20,
    position: "absolute"
  };

  const labelStyle: React.CSSProperties = {
    cursor: "pointer",
    paddingLeft: 20,
    paddingRight: 10
  };

  return (
    <label style={labelStyle}>
      <input
        type="checkbox"
        name={name}
        value={value}
        disabled={disabled}
        defaultChecked={checked}
        onChange={onChange}
        style={inputStyle}
      />
      {children}
    </label>
  );
};

export default Checkbox;
