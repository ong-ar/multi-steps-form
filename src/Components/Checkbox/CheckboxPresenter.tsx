import * as React from "react";

interface IProps {
  name: string;
  disabled?: boolean;
  checked?: boolean;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const CheckboxPresenter: React.SFC<IProps> = ({
  name,
  disabled = false,
  checked = false,
  onChange,
  children
}) => {
  const inputStyle: React.CSSProperties = {
    marginLeft: -20,
    position: "absolute"
  };

  const labelStyle: React.CSSProperties = {
    cursor: "pointer",
    paddingLeft: 20
  };

  return (
    <label style={labelStyle}>
      <input
        type="checkbox"
        name={name}
        disabled={disabled}
        checked={checked}
        onChange={onChange}
        style={inputStyle}
      />
      {children}
    </label>
  );
};

export default CheckboxPresenter;
