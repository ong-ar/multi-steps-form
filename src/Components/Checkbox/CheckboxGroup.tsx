import * as React from "react";
import Checkbox from "./Checkbox";

interface IProps {
  name?: string;
  defaultValues?: string[];
  onChange?: (checkedValue: string, groupId?: string | number) => void;
  groupId?: string | number; // 부모노드에서 ID가 필요할 경우 사용
}

const CheckboxGroup: React.FC<IProps> = ({
  name,
  defaultValues,
  onChange,
  children,
  groupId
}) => {
  const [values, setValues] = React.useState<string[]>(defaultValues || []);

  // Checkbox Change handler
  const onCheckboxChange: React.ChangeEventHandler<HTMLInputElement> = e => {
    let changedValues: string[] = [];
    if (e.target.checked) {
      changedValues = [...values, e.target.value];
    } else {
      changedValues = values.filter(value => value !== e.target.value);
    }

    // set state
    setValues(changedValues);

    // CheckboxGroup onChange
    if (onChange) {
      onChange(changedValues.join(), groupId);
    }
  };

  const checkboxGroup = {
    name,
    onChange: onCheckboxChange
  };

  // children 에 필요한 값 추가
  const customedChildren = React.Children.map(
    children,
    (child: React.ReactElement) => {
      if (child && child.type === Checkbox) {
        // defaultValues 적용
        if (defaultValues && defaultValues.includes(child.props.value)) {
          return React.cloneElement(child, {
            ...checkboxGroup,
            defaultChecked: true
          });
        } else {
          return React.cloneElement(child, {
            ...checkboxGroup,
            defaultChecked: false
          });
        }
      } else {
        return child;
      }
    }
  );
  return <div>{customedChildren}</div>;
};

export default CheckboxGroup;
