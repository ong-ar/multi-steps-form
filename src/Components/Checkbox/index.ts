import CheckboxPresenter from "./CheckboxPresenter";
export default CheckboxPresenter;
