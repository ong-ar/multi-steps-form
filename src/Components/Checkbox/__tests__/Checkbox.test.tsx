import * as React from "react";
import { shallow } from "enzyme";

import Checkbox from "../Checkbox";

// checkbox
it("Checkbox", () => {
  const wrapper = shallow(<Checkbox />);
  expect(wrapper).toMatchSnapshot();
});

// props: name 체크
it("Checkbox:name", () => {
  const wrapper = shallow(<Checkbox name="checkbox" />);
  expect(wrapper.find("input").prop("name")).toEqual("checkbox");
});

// props: value 체크
it("Checkbox:value", () => {
  const wrapper = shallow(<Checkbox value="checkbox" />);
  expect(wrapper.find("input").prop("value")).toEqual("checkbox");
});

// props: disabled 체크
it("Checkbox:disabled", () => {
  const wrapper = shallow(<Checkbox disabled={true} />);
  expect(wrapper.find("input").prop("disabled")).toEqual(true);
});

// props: checked 체크
it("Checkbox:checked", () => {
  const wrapper = shallow(<Checkbox checked={false} />);
  expect(wrapper.find("input").prop("checked")).toEqual(false);
});

// props: defaultChecked 체크
it("Checkbox:defaultChecked", () => {
  const wrapper = shallow(<Checkbox defaultChecked={true} />);
  expect(wrapper.find("input").prop("defaultChecked")).toEqual(true);
});
