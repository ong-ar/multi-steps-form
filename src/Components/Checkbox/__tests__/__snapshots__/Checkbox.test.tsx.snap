// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Checkbox />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <label
        style={
                Object {
                        "cursor": "pointer",
                        "paddingLeft": 20,
                        "paddingRight": 10,
                      }
        }
>
        <input
                checked={undefined}
                defaultChecked={undefined}
                disabled={undefined}
                name={undefined}
                onChange={undefined}
                style={
                        Object {
                                "marginLeft": -20,
                                "position": "absolute",
                              }
                }
                type="checkbox"
                value={undefined}
        />
</label>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked={undefined}
            defaultChecked={undefined}
            disabled={undefined}
            name={undefined}
            onChange={undefined}
            style={
                        Object {
                                    "marginLeft": -20,
                                    "position": "absolute",
                                  }
            }
            type="checkbox"
            value={undefined}
/>,
          undefined,
        ],
        "style": Object {
          "cursor": "pointer",
          "paddingLeft": 20,
          "paddingRight": 10,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "checked": undefined,
            "defaultChecked": undefined,
            "disabled": undefined,
            "name": undefined,
            "onChange": undefined,
            "style": Object {
              "marginLeft": -20,
              "position": "absolute",
            },
            "type": "checkbox",
            "value": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        undefined,
      ],
      "type": "label",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <label
          style={
                    Object {
                              "cursor": "pointer",
                              "paddingLeft": 20,
                              "paddingRight": 10,
                            }
          }
>
          <input
                    checked={undefined}
                    defaultChecked={undefined}
                    disabled={undefined}
                    name={undefined}
                    onChange={undefined}
                    style={
                              Object {
                                        "marginLeft": -20,
                                        "position": "absolute",
                                      }
                    }
                    type="checkbox"
                    value={undefined}
          />
</label>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              checked={undefined}
              defaultChecked={undefined}
              disabled={undefined}
              name={undefined}
              onChange={undefined}
              style={
                            Object {
                                          "marginLeft": -20,
                                          "position": "absolute",
                                        }
              }
              type="checkbox"
              value={undefined}
/>,
            undefined,
          ],
          "style": Object {
            "cursor": "pointer",
            "paddingLeft": 20,
            "paddingRight": 10,
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "checked": undefined,
              "defaultChecked": undefined,
              "disabled": undefined,
              "name": undefined,
              "onChange": undefined,
              "style": Object {
                "marginLeft": -20,
                "position": "absolute",
              },
              "type": "checkbox",
              "value": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          undefined,
        ],
        "type": "label",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
